CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityLinkInventoryBase.h

```diff
@@ -0,0 +1,233 @@
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkInventoryTracker {
+	public:
+
+		FPPLinkInventoryTracker() {};
+		~FPPLinkInventoryTracker() {};
+		
+		FPPLinkInventoryBase<LinkType> linkInv;
+	};
+
+
+	template <class LinkType>
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkInventoryBase {
+
+	public:
+
+		typedef FPPLinkListBase<LinkType> FPPLinkList;
+		typedef FPPLinkInventoryBase<LinkType> FPPInventory_t;
+
+		typedef std::map<const FPPLinkID, LinkType*> linkInventory_t;
+		typedef typename linkInventory_t::iterator linkInventoryItr_t;
+		typedef std::pair<const FPPLinkID, LinkType*> linkInventoryPair_t;
+
+	protected:
+
+		Potts3D* potts;
+		BasicClassAccessor<FocalPointPlasticityTracker>* focalPointPlasticityTrackerAccessor;
+		BasicClassAccessor<FPPLinkInventoryTracker<LinkType> >* cellLinkInventoryTrackerAccessor;
+
+		linkInventory_t linkInventory;
+
+		LinkType* getLinkById(const FPPLinkID _id) {
+			linkInventoryItr_t itr = linkInventory.find(_id);
+			if (itr != linkInventory.end()) return itr->second;
+			return (LinkType*)(0);
+		}
+		const FPPLinkID getLinkId(LinkType* _link) { return FPPLinkID(_link->getId0(), _link->getId1()); }
+
+		// Add a link without any additional internal work
+		void addLinkNoChain(LinkType* _link) {
+			linkInventory.insert(linkInventoryPair_t(getLinkId(_link), _link));
+		}
+		// Add tracker data to cells
+		void addTrackerData(LinkType* _link) {
+			CellG* cell0 = _link->getObj0();
+			CellG* cell1 = _link->getObj1();
+			if (cell0) {
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell0);
+				getFPPTrackerDataSet(cell0).insert(fppd);
+			}
+			if (cell1) {
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell1);
+				getFPPTrackerDataSet(cell1).insert(fppd);
+			}
+		}
+		// Remove a link without any additional internal work
+		void removeLinkNoChain(LinkType* _link) {
+			const FPPLinkID linkId = getLinkId(_link);
+			linkInventoryItr_t itr = linkInventory.find(linkId);
+			if (itr != linkInventory.end()) linkInventory.erase(linkId);
+		}
+		// Remove tracker data from cells
+		void removeTrackerData(LinkType* _link) {
+			FPPTrackerDataSet::iterator itr;
+			CellG* cell0 = const_cast<CellG*>(_link->getObj0());
+			CellG* cell1 = const_cast<CellG*>(_link->getObj1());
+			if (cell0) {
+				FPPTrackerDataSet &fppdSet = getFPPTrackerDataSet(cell0);
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell0);
+				itr = fppdSet.find(fppd);
+				if (itr != fppdSet.end()) fppdSet.erase(fppd);
+			}
+			if (cell1) {
+				FPPTrackerDataSet &fppdSet = getFPPTrackerDataSet(cell1);
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell1);
+				itr = fppdSet.find(fppd);
+				if (itr != fppdSet.end()) fppdSet.erase(fppd);
+			}
+		}
+
+		FPPInventory_t* getCellLinkInventory(CellG* _cell) { 
+			return &cellLinkInventoryTrackerAccessor->get(_cell->extraAttribPtr)->linkInv;
+		}
+
+		CellG* getCellGFromConst(const CellG* _cell) {
```

Comment:
In general std::const_cast<CellG*> should be used - this si standard C++ way of dealing with constantness. I also noticed a lot of overloaded constructors in one of the class that differ only by the fact that some of the args are constants and some are not. I would avoid this typo of overloads because they make code hard to read

Reply:
I would remove this function completely. It does not belong to inventory

Reply:
Implemented in https://github.com/CompuCell3D/CompuCell3D/pull/361/commits/1305c536c83539572ee8f043066638c4d3c5e949, https://github.com/CompuCell3D/CompuCell3D/pull/361/commits/5e10487f5061e77cf90a34f83bd8d9f5231628a0, and https://github.com/CompuCell3D/CompuCell3D/pull/361/commits/9e3315490aa05a7c7cf8c66012cfe12c3ba4e0c6